src/csharp/Grpc.Core.Api/Marshaller.cs

```diff
@@ -41,12 +42,51 @@ public Marshaller(Func<T, byte[]> serializer, Func<byte[], T> deserializer)
         {
             this.serializer = GrpcPreconditions.CheckNotNull(serializer, nameof(serializer));
             this.deserializer = GrpcPreconditions.CheckNotNull(deserializer, nameof(deserializer));
+
             // contextual serialization/deserialization is emulated to make the marshaller
             // usable with the grpc library (required for backward compatibility).
             this.contextualSerializer = EmulateContextualSerializer;
-            this.contextualDeserializer = EmulateContextualDeserializer;
+
+            if (!TryFindArraySegmentDeserializer(deserializer, out this.contextualDeserializer))
+            {
+                this.contextualDeserializer = EmulateContextualDeserializer;
+            }
         }
 
+        /// <summary>
+        /// Given a deserializer of the form foo.Bar(byte[]), see if we can find a similar foo.Bar(byte[], int, int); if so,
+        /// prefer that; this allows the existing protoc code-gen to be identified to use pooled buffers
+        /// </summary>
+        private static bool TryFindArraySegmentDeserializer(Func<byte[], T> deserializer, out Func<DeserializationContext, T> contextualDeserializer)
```

Comment:
+1 to the idea of trying to make use the `foo.Bar(byte[], int, int)` parser without requiring any codegen changes,
I think that is something that's worth investigating. But we also need to give this a bit more thought so that it doesn't have any unwanted consequences (e.g. protobuf is not the only serialization format supported and we don't know what this code would do when other serialization formats are being used).

Reply:
We could perhaps further restrict it to where the first parameter matches name and the second parameters are named `offset` and `count`? I think that's pretty unambiguous about the intent...